### PR TITLE
Improve code around file system watching

### PR DESCRIPTION
**Pull request by wolfs**
The major change is that there isn't a `WatchingVirtualFileSystem` any more. There now is a watch handler, which takes care of the orchestration of the watchers.

### CODE REVIEW

File: subprojects/core/src/main/java/org/gradle/internal/service/scopes/VirtualFileSystemServices.java [hunk 1]
```diff
@@ -173,30 +175,52 @@ FileHasher createCachingFileHasher(StringInterner stringInterner, CrossBuildFile
             return fileHasher;
         }
 
-        WatchingAwareVirtualFileSystem createVirtualFileSystem(
-            GlobalCacheLocations globalCacheLocations,
+        RecentlyCapturedSnapshots createRecentlyCapturedSnapshots(ListenerManager listenerManager) {
```
Comment (lptr) [inline subprojects/core/src/main/java/org/gradle/internal/service/scopes/VirtualFileSystemServices.java:178]: Would be nice to have a more precise definition here than "recently".

File: subprojects/file-watching/src/main/java/org/gradle/internal/watch/vfs/impl/DefaultFileSystemWatchingHandler.java [hunk 2]
```diff
@@ -44,24 +42,20 @@
...
- */
-public class WatchingVirtualFileSystem extends AbstractDelegatingVirtualFileSystem implements WatchingAwareVirtualFileSystem, Closeable {
-    private static final Logger LOGGER = LoggerFactory.getLogger(WatchingVirtualFileSystem.class);
+public class DefaultFileSystemWatchingHandler implements FileSystemWatchingHandler, Closeable {
+    private static final Logger LOGGER = LoggerFactory.getLogger(DefaultFileSystemWatchingHandler.class);
     private static final String FILE_WATCHING_ERROR_MESSAGE_DURING_BUILD = "Unable to watch the file system for changes";
     private static final String FILE_WATCHING_ERROR_MESSAGE_AT_END_OF_BUILD = "Gradle was unable to watch the file system for changes";
 
     private final FileWatcherRegistryFactory watcherRegistryFactory;
+    private final AbstractVirtualFileSystem virtualFileSystem;
```
Comment (lptr) [inline subprojects/file-watching/src/main/java/org/gradle/internal/watch/vfs/impl/DefaultFileSystemWatchingHandler.java:54]: There is a code smell here... `AbstractVirtualFileSystem` should be an interface instead. Maybe what we need is a `VFS.update(UpdateFunction)` method.

File: subprojects/launcher/src/main/java/org/gradle/tooling/internal/provider/ContinuousBuildActionExecuter.java [hunk 3]
```diff
@@ -77,6 +78,9 @@ public ContinuousBuildActionExecuter(
 
     @Override
     public BuildActionResult execute(BuildAction action, BuildRequestContext requestContext, final BuildActionParameters actionParameters, ServiceRegistry buildSessionScopeServices) {
+        // Ensure that file system watching is enabled
+        // TODO: Use the handler later on to do things with continuous build
+        buildSessionScopeServices.get(FileSystemWatchingHandler.class);
```
Comment (lptr) [inline subprojects/launcher/src/main/java/org/gradle/tooling/internal/provider/ContinuousBuildActionExecuter.java:83]: Why is this needed?

File: subprojects/snapshots/src/main/java/org/gradle/internal/vfs/impl/DefaultVirtualFileSystem.java [hunk 4]
```diff
@@ -178,6 +187,7 @@ private CompleteFileSystemLocationSnapshot readLocation(String location) {
 
     @Override
     public void update(Iterable<String> locations, Runnable action) {
+        recentlyCreatedSnapshotsListener.snapshotsCreated(locations);
```
Comment (lptr) [inline subprojects/snapshots/src/main/java/org/gradle/internal/vfs/impl/DefaultVirtualFileSystem.java:190]: Are they created or updated?

File: subprojects/launcher/src/main/java/org/gradle/tooling/internal/provider/ContinuousBuildActionExecuter.java (hunk 3 above)
Comment (wolfs) [reply subprojects/launcher/src/main/java/org/gradle/tooling/internal/provider/ContinuousBuildActionExecuter.java:83]: See https://gradle.slack.com/archives/CDYR952SH/p1594659521457500

I'll add a clarifying comment.

File: subprojects/snapshots/src/main/java/org/gradle/internal/vfs/impl/DefaultVirtualFileSystem.java (hunk 4 above)
Comment (wolfs) [reply subprojects/snapshots/src/main/java/org/gradle/internal/vfs/impl/DefaultVirtualFileSystem.java:190]: Currently, they are invalidated. At some point, we wanted to store the recently created things in there: https://github.com/gradle/gradle/issues/12761

Let's do renaming together.

File: subprojects/file-watching/src/main/java/org/gradle/internal/watch/vfs/impl/DefaultFileSystemWatchingHandler.java (hunk 2 above)
Comment (wolfs) [reply subprojects/file-watching/src/main/java/org/gradle/internal/watch/vfs/impl/DefaultFileSystemWatchingHandler.java:54]: That sounds like `VirtualFileSystem` becoming an `AtomicSnapshotHierarchyReference`. That isn't necessarily bad. We also could separate the `AtomicSnapshotHierarchyReference` from the VFS making it its own first class object. The snapshot hierarchy is the data in the VFS, while `VirtualFileSystem` allows querying the actual file system, using the backing `SnapshotHierarchy`.

`FileSystemWatchingHandler` operates only on the snapshot hierarchy, publishing changes to it to the watcher registry.

I'll try something out. Let's look at this together tomorrow as well.

File: subprojects/launcher/src/main/java/org/gradle/tooling/internal/provider/ContinuousBuildActionExecuter.java (hunk 3 above)
Comment (wolfs) [reply subprojects/launcher/src/main/java/org/gradle/tooling/internal/provider/ContinuousBuildActionExecuter.java]: Added.

File: subprojects/file-watching/src/main/java/org/gradle/internal/watch/vfs/impl/DefaultFileSystemWatchingHandler.java (hunk 2 above)
Comment (wolfs) [reply subprojects/file-watching/src/main/java/org/gradle/internal/watch/vfs/impl/DefaultFileSystemWatchingHandler.java]: `FileSystemWatchingHandler` now only uses the `AtomicSnapshotHierarchyReference` and `AbstractVirtualFileSystem` is gone.

File: subprojects/file-watching/src/main/java/org/gradle/internal/watch/vfs/impl/DefaultFileSystemWatchingHandler.java [hunk 5]
```diff
@@ -73,32 +67,30 @@
...
         this.watchFilter = watchFilter;
         this.daemonDocumentationIndex = daemonDocumentationIndex;
+        this.locationsUpdatedByCurrentBuild = locationsUpdatedByCurrentBuild;
     }
 
     @Override
     public void afterBuildStarted(boolean watchingEnabled) {
         reasonForNotWatchingFiles = null;
-        getRoot().update(currentRoot -> {
+        snapshotHierarchyReference.update(currentRoot -> {
```
Comment (lptr) [inline subprojects/file-watching/src/main/java/org/gradle/internal/watch/vfs/impl/DefaultFileSystemWatchingHandler.java:89]: I think this name is confusing. What we are updating is the `root`, or the reference to the root. I think this would read a lot nicer:

```suggestion
        root.update(currentRoot -> {
```

Comment (wolfs) [reply subprojects/file-watching/src/main/java/org/gradle/internal/watch/vfs/impl/DefaultFileSystemWatchingHandler.java]: I renamed `snapshotHierarchyReference` to `root`.
